object_store/src/azure/credential.rs

```diff
@@ -0,0 +1,254 @@
+                    .header(
+                        AUTHORIZATION,
+                        HeaderValue::from_str(signature.as_str()).unwrap(),
+                    );
+            }
+            AzureCredential::AuthorizationToken(token) => {
+                self = self.header(AUTHORIZATION, token);
+            }
+            AzureCredential::SASToken(query_pairs) => {
+                self = self.query(&query_pairs);
+            }
+        };
+
+        self
+    }
+}
+
+// generate signed key for authorization via access keys
+fn generate_authorization(
+    h: &HeaderMap,
+    u: &Url,
+    method: &Method,
+    account: &str,
+    key: &str,
+) -> String {
+    let str_to_sign = string_to_sign(h, u, method, account);
+    let auth = hmac_sha256(base64::decode(key).unwrap(), &str_to_sign);
+    format!("SharedKey {}:{}", account, base64::encode(auth))
+}
+
+fn add_if_exists<'a>(h: &'a HeaderMap, key: &HeaderName) -> &'a str {
+    h.get(key)
+        .map(|s| s.to_str())
+        .transpose()
+        .ok()
+        .unwrap_or(Some(""))
+        .unwrap_or_default()
+}
+
+fn string_to_sign(h: &HeaderMap, u: &Url, method: &Method, account: &str) -> String {
+    // content length must only be specified if != 0
+    // this is valid from 2015-02-21
+    let content_length = h
+        .get(&CONTENT_LENGTH)
+        .map(|s| s.to_str())
+        .transpose()
+        .ok()
+        .unwrap_or(Some(""))
+        .filter(|&v| v != "0")
+        .unwrap_or_default();
+    format!(
+        "{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}{}",
+        method.as_ref(),
+        add_if_exists(h, &CONTENT_ENCODING),
+        add_if_exists(h, &CONTENT_LANGUAGE),
+        content_length,
+        add_if_exists(h, &CONTENT_MD5),
+        add_if_exists(h, &CONTENT_TYPE),
+        add_if_exists(h, &DATE),
+        add_if_exists(h, &IF_MODIFIED_SINCE),
+        add_if_exists(h, &IF_MATCH),
+        add_if_exists(h, &IF_NONE_MATCH),
+        add_if_exists(h, &IF_UNMODIFIED_SINCE),
+        add_if_exists(h, &RANGE),
+        canonicalize_header(h),
+        canonicalized_resource(account, u)
+    )
+}
+
+fn canonicalize_header(headers: &HeaderMap) -> String {
+    let mut names = headers
+        .iter()
+        .filter_map(|(k, _)| {
+            (k.as_str().starts_with("x-ms"))
+                // TODO remove unwraps
+                .then(|| (k.as_str(), headers.get(k).unwrap().to_str().unwrap()))
+        })
+        .collect::<Vec<_>>();
+    names.sort_unstable();
+
+    let mut result = String::new();
+    for (name, value) in names {
+        result = format!("{result}{name}:{value}\n");
+    }
+    result
+}
+
+fn canonicalized_resource(account: &str, uri: &Url) -> String {
+    let mut can_res: String = String::new();
+    can_res += "/";
+    can_res += account;
+
+    for p in uri.path_segments().into_iter().flatten() {
+        can_res.push('/');
+        can_res.push_str(p);
+    }
+    can_res += "\n";
```

Comment:
```suggestion
    can_res.push('\n');
```

Not actually sure if there is a difference